### PR TITLE
new stuff for kenny from Newstuffforkenny

### PR DESCRIPTION
**Pull request by kendama**

### CODE REVIEW

File: activities2Graph.py [hunk 1]
```diff
@@ -51,34 +51,38 @@ def idGenerator(start=0):
...
+            #retrieve previous versions
+            if int(versionNumber) > 1:
+                for version in range(1,int(versionNumber)):
+                    ent['_id'] = newId.next()
+                    entityDict['%s.%s' %(ent['synId'],version)] = ent 
+                    logging.info('Getting previous version of entity (%i): %s.%i' %(ent['_id'],
+                                                 ent['synId'], version))
+        return entityDict
+    except synapseclient.exceptions.SynapseHTTPError:
+        return entityDict
```
Comment (kdaily) [inline activities2Graph.py]: Move try inside `for ent in query` loop - return both `entityDict` and a list of bad entities caught by the exception?

File: activities2Graph.py [hunk 2]
```diff
@@ -51,34 +51,38 @@ def idGenerator(start=0):
...
+            ent['_id'] = newId.next()
+            ent['synId'] = ent.pop('id')
+
+            versionNumber = ent['versionNumber']
+            entityDict['%s.%s' %(ent['synId'],versionNumber)] = ent 
+            logging.info('Getting entity (%i): %s.%s' %(ent['_id'], ent['synId'],
+                                                 ent['versionNumber']))
+            #retrieve previous versions
+            if int(versionNumber) > 1:
+                for version in range(1,int(versionNumber)):
```
Comment (kdaily) [inline activities2Graph.py]: No guarantee that all versions between 1 and the current version exist. Use [REST call](http://hud.rel.rest.doc.sagebase.org.s3-website-us-east-1.amazonaws.com/GET/entity/id/version.html) to get all versions of a file.

File: activities2Graph.py [hunk 3]
```diff
@@ -51,34 +51,38 @@ def idGenerator(start=0):
...
-                                             ent['synId'], version))
-    return entityDict
+    global newId
+    try:
+        logging.info('Getting and formatting all entities with the inputted projectId')
+        query = syn.chunkedQuery('select * from entity where projectId =="%s"' %projectId)
+        entityDict = dict()
+        for ent in query:
+            if ent['entity.nodeType'] in [1,2,3]: 
+	    #Remove containers by ignoring layers, projects, and previews
```
Comment (kdaily) [inline activities2Graph.py]: move comment up, abstract the list `[1,2,3]` to a global var so it's defined and can be modified if necessary.

File: activities2Graph.py [hunk 4]
```diff
@@ -51,34 +51,38 @@ def idGenerator(start=0):
...
-                                             ent['versionNumber']))
-        #retrieve previous versions
-        if int(versionNumber) > 1:
-            for version in range(1,int(versionNumber)):
-                ent['_id'] = newId.next()
-                entityDict['%s.%s' %(ent['synId'],version)] = ent 
-                logging.info('Getting previous version of entity (%i): %s.%i' %(ent['_id'],
-                                             ent['synId'], version))
-    return entityDict
+    global newId
```
Comment (kdaily) [inline activities2Graph.py]: use local var, pass into function.

File: activities2Graph.py [hunk 5]
```diff
@@ -92,6 +96,7 @@ def safeGetActivity(entity):
 
 def cleanUpActivities(activities):
     '''remove all activity-less entities'''
+    global newId
```
Comment (kdaily) [inline activities2Graph.py]: pass in `newId` generator

File: activities2Graph.py [hunk 6]
```diff
@@ -120,7 +125,10 @@ def buildEdgesfromActivities(nodes, activities):
             new_nodes[activity['synId']]  = activity
             #Add input relationships
             for used in activity['used']:
-                edges = addNodesandEdges(used, nodes, activity, edges)
+                try:
```
Comment (kdaily) [inline activities2Graph.py]: move try/catch inside of `addNodesandEdges` - around specific place(s) where there are errors. consider logging errors to a static file as well as the console (stderr)

File: convertSynapse2Graph.py [hunk 7]
```diff
@@ -48,47 +49,52 @@ def idGenerator(start=0):
...
-            item = ent.pop(key)
-            ent[new_key] = item[0] if (type(item) is list and len(item)>0) else item
-        ent['_type']='vertex'
-        ent['_id'] = newId.next()
-        ent['synId'] = ent.pop('id')
-
-        versionNumber = ent['versionNumber']
-        entityDict['%s.%s' %(ent['synId'],versionNumber)] = ent 
-        logging.info('Getting entity (%i): %s.%s' %(ent['_id'], ent['synId'],
+    global newId
```
Comment (kdaily) [inline convertSynapse2Graph.py]: pass in instead of global

File: convertSynapse2Graph.py [hunk 8]
```diff
@@ -48,47 +49,52 @@ def idGenerator(start=0):
...
+    try:
```
Comment (kdaily) [inline convertSynapse2Graph.py]: move try inside of for loop

File: convertSynapse2Graph.py [hunk 9]
```diff
@@ -48,47 +49,52 @@ def idGenerator(start=0):
 
 def getEntities(projectId):
```
Comment (kdaily) [inline convertSynapse2Graph.py]: merge with `getEntities` that is in `activities2Graph.py` (or move to a shared library file and call the same one)

File: convertSynapse2Graph.py [hunk 10]
```diff
@@ -170,13 +178,21 @@ def addNodesandEdges(used, nodes, activity, edges):
     logger = logging.getLogger()
     logger.setLevel(logging.INFO)
 
-    p = mp.Pool(7)
+    parser = argparse.ArgumentParser(description=
+                'Creates a json file based on provenance for all Synapse projects')
+    parser.add_argument('--p', type=int, default=6, help='Specify the pool size for the multiprocessing module')
```
Comment (kdaily) [inline convertSynapse2Graph.py]: default to 2 or 4?

File: convertSynapse2Graph.py [hunk 11]
```diff
@@ -170,13 +178,21 @@ def addNodesandEdges(used, nodes, activity, edges):
...
+    args = parser.parse_args()
+
+    p = mp.Pool(args.p)
     projects = syn.chunkedQuery("select id from project")
     nodes = dict()
 
     for proj in projects:
+        if proj in ['syn582072', 'syn3218329', 'syn2044761', 'syn2351328', 'syn1450028']:
```
Comment (kdaily) [inline convertSynapse2Graph.py]: define this at the top (or allow parameter for it)

File: convertSynapse2Graph.py [hunk 12]
```diff
@@ -185,6 +201,6 @@ def addNodesandEdges(used, nodes, activity, edges):
                                                                         float(len(nodes))/len(activities))
     edges = buildEdgesfromActivities(nodes, activities)
     logging.info('I have  %i nodes and %i edges' %(len(nodes), len(edges)))
-    with open('graph.json', 'w') as fp:
+    with open('graphSynapse.json', 'w') as fp:
```
Comment (kdaily) [inline convertSynapse2Graph.py]: allow user to specify name of output file (or just use stdout)

File: examples/cypher_queries.txt [hunk 13]
```diff
@@ -0,0 +1,114 @@
+## Useful Cypher queries for Neo4j Database
```
Comment (kdaily) [inline examples/cypher_queries.txt]: rename as `cypher_queries.md` so Github renders as Markdown.

File: load2Neo4jDB.py [hunk 14]
```diff
@@ -70,12 +78,14 @@ def json2neo4j(jsonfilename, graph):
     aQuery = aQuery % nodes.name
     bQuery = bQuery % nodes.name
     cQuery = cQuery % edges.name
+    dQuery = dQuery % edges.name
```
Comment (kdaily) [inline load2Neo4jDB.py]: give more descriptive names for the queries

File: load2Neo4jDB.py [hunk 15]
```diff
@@ -70,12 +78,14 @@ def json2neo4j(jsonfilename, graph):
...
 
     # Send Cypher query
     logging.info('Loading data from CSV file(s) to Neo4j')
     graph.run(aQuery)
     graph.run(bQuery)
     graph.run(cQuery)
+    graph.run(dQuery)
```
Comment (kdaily) [inline load2Neo4jDB.py]: turn into a loop over all queries to run with `graph.run`.

File: load2Neo4j_deprecated.py [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+import csv
+import pandas as pd
+import sys
+import os
+from py2neo import Graph, authenticate
+
+gc.collect()
+# Connect to graph and add constraints.
+print 'Connecting to Neo4j and authenticating user credentials'
+authenticate("ec2-54-91-28-8.compute-1.amazonaws.com:7474", "neo4j", "sage2016")
```
Comment (kdaily) [inline load2Neo4j_deprecated.py]: remove this and restart machine

File: load2Neo4j_deprecated.py [hunk 17]
```diff
@@ -0,0 +1,106 @@
+import gc
```
Comment (kdaily) [inline load2Neo4j_deprecated.py]: remove whole file - use diff version of `load2Neo4j.py`.

**Review by kdaily — CHANGES_REQUESTED** [review]
Consolidate all fxns duplicated across scripts. Remove printing to stdout (print to stderr using `sys.stderr.write`)

File: activities2Graph.py (hunk 3 above)
Comment (kendama) [inline activities2Graph.py]: doneish

File: activities2Graph.py (hunk 4 above)
Comment (kendama) [inline activities2Graph.py]: done

File: activities2Graph.py (hunk 5 above)
Comment (kendama) [inline activities2Graph.py]: done

File: convertSynapse2Graph.py (hunk 9 above)
Comment (kendama) [inline convertSynapse2Graph.py]: done

File: convertSynapse2Graph.py (hunk 7 above)
Comment (kendama) [inline convertSynapse2Graph.py]: done

File: convertSynapse2Graph.py (hunk 10 above)
Comment (kendama) [inline convertSynapse2Graph.py]: used 4

File: convertSynapse2Graph.py (hunk 11 above)
Comment (kendama) [inline convertSynapse2Graph.py]: done

File: examples/cypher_queries.txt (hunk 13 above)
Comment (kendama) [inline examples/cypher_queries.txt]: done

File: activities2Graph.py (hunk 1 above)
Comment (kendama) [inline activities2Graph.py]: done

File: activities2Graph.py (hunk 6 above)
Comment (kendama) [inline activities2Graph.py]: done

File: convertSynapse2Graph.py (hunk 12 above)
Comment (kendama) [inline convertSynapse2Graph.py]: finito

File: load2Neo4jDB.py (hunk 14 above)
Comment (kendama) [inline load2Neo4jDB.py]: done

File: load2Neo4jDB.py (hunk 15 above)
Comment (kendama) [inline load2Neo4jDB.py]: done

**Comment by kdaily** [comment]
I see responses to most changes (hooray!) but don't see any changes to the code from what we reviewed? ¿Que paso?

File: convertSynapse2Graph.py [hunk 18]
```diff
@@ -66,21 +72,30 @@ def getEntities(projectId):
...
+                    ent['synId'] = synId
+                    ent['versionNumber'] = version
                     entityDict['%s.%s' %(ent['synId'],version)] = ent 
                     logging.info('Getting previous version of entity (%i): %s.%i' %(ent['_id'],
                                                  ent['synId'], version))
-        return entityDict
-    except synapseclient.exceptions.SynapseHTTPError:
-        return entityDict
+        except synapseclient.exceptions.SynapseHTTPError:
+            continue
```
Comment (kdaily) [inline convertSynapse2Graph.py]: add logging statement here with error and synapse ID of entity

File: convertSynapse2Graph.py [hunk 19]
```diff
@@ -144,21 +155,24 @@ def buildEdgesfromActivities(nodes, activities):
...
+        try:
+            targetId = '%s.%s' %(used['reference']['targetId'],
+                                 used['reference'].get('targetVersionNumber'))
+            if targetId not in nodes:
+                nodes[targetId] = { '_id': newId.next(),
+                                    '_type': 'vertex',
+                                    'synId' : used['reference']['targetId'],
+                                    'versionNumber': used['reference'].get('targetVersionNumber')}
+        except KeyError as e:
+            sys.stderr.write('Skipping current relationship retrieval due to targetId KeyError')
```
Comment (kdaily) [inline convertSynapse2Graph.py]: add id of activity in output

File: convertSynapse2Graph.py [hunk 20]
```diff
@@ -66,21 +72,30 @@ def getEntities(projectId):
...
-
+            synId = ent['synId']
             versionNumber = ent['versionNumber']
             entityDict['%s.%s' %(ent['synId'],versionNumber)] = ent 
             logging.info('Getting entity (%i): %s.%s' %(ent['_id'], ent['synId'],
                                              ent['versionNumber']))
             #retrieve previous versions
             if int(versionNumber) > 1:
                 for version in range(1,int(versionNumber)):
+                    ent = syn.restGET('/entity/%s/version/%s' %(ent['synId'],version))
```
Comment (kdaily) [inline convertSynapse2Graph.py]: `ent = syn.get(, version=?)`

**Review by kdaily — CHANGES_REQUESTED** [review]

**Review by kdaily — APPROVED** [review]

**Pull request merged**